proposals/handler.go

```diff
@@ -519,6 +523,88 @@ func (h *Handler) checkBallotSyntacticValidity(
 	return decoded, nil
 }
 
+func (h *Handler) getActiveSetWeight(ctx context.Context, id types.Hash32) (uint64, error) {
+	h.weightCalcLock.Lock()
+	totalWeight, exists := h.activeSets.Get(id)
+	if exists {
+		h.weightCalcLock.Unlock()
+		return totalWeight, nil
+	}
+
+	var ch chan uint64
+	chs, exists := h.pendingWeightCalc[id]
+	if exists {
+		// The calculation is running or the activeset is being fetched,
+		// subscribe.
+		// Avoid any blocking on the channel by making it buffered, also so that
+		// we don't have to wait on it in case the context is canceled
+		ch = make(chan uint64, 1)
+		h.pendingWeightCalc[id] = append(chs, ch)
+	} else {
+		// mark calculation as running
+		h.pendingWeightCalc[id] = nil
+	}
+	h.weightCalcLock.Unlock()
+
+	if exists {
+		// need to wait for the calculation which is already running to finish
+		select {
+		case <-ctx.Done():
+			return 0, ctx.Err()
+		case totalWeight, ok := <-ch:
+			if !ok {
+				// Channel closed, fetch / calculation failed.
+				// The actual error will be logged by the initiator of the
+				// initial fetch / calculation, let's not make an
+				// impression it happened multiple times and use a simpler
+				// message
+				return totalWeight, errors.New("error getting activeset weight")
+			}
+			return totalWeight, nil
+		}
```

Comment:
There could be fewer branches by restructuring the code:
```suggestion
	if exists {
		// The calculation is running or the activeset is being fetched,
		// subscribe.
		// Avoid any blocking on the channel by making it buffered, also so that
		// we don't have to wait on it in case the context is canceled
		ch = make(chan uint64, 1)
		h.pendingWeightCalc[id] = append(chs, ch)
		// need to wait for the calculation which is already running to finish
		select {
		case <-ctx.Done():
			return 0, ctx.Err()
		case totalWeight, ok := <-ch:
			if !ok {
				// Channel closed, fetch / calculation failed.
				// The actual error will be logged by the initiator of the
				// initial fetch / calculation, let's not make an
				// impression it happened multiple times and use a simpler
				// message
				return totalWeight, errors.New("error getting activeset weight")
			}
			return totalWeight, nil
	}
	// mark calculation as running
	h.pendingWeightCalc[id] = nil
	h.weightCalcLock.Unlock()
```

Reply:
This would block on the channel while holding the lock on _all_ activeset processing in the `Handler` (not just for this activeset)

Reply:
Sorry, forgot to unlock. Updated the proposal.

Reply:
Still problematic perhaps due to missing `else` (mishandling or running calculation, double unlock, missing `}`)
;)
But I understand what you mean, so I tried to simplify this part